hibernate-core/src/main/java/org/hibernate/internal/SessionImpl.java

```diff
@@ -2063,14 +2063,20 @@ private SharedSessionBuilderImpl(SessionImpl session) {
 
 		@Override
 		public SharedSessionBuilderImpl tenantIdentifier(String tenantIdentifier) {
-			// todo : is this always true?  Or just in the case of sharing JDBC resources?
-			throw new SessionException( "Cannot redefine tenant identifier on child session" );
+			if ( !session.getSessionFactory().getSessionFactoryOptions().isMultiTenancyEnabled() ) {
+				throw new SessionException( "Cannot redefine tenant identifier on child session" );
+			}
+			session.tenantIdentifier = session.getSessionFactory().getTenantIdentifierJavaType().toString( tenantIdentifier );
```

Comment:
```suggestion
			this.tenantIdentifier = session.getSessionFactory().getTenantIdentifierJavaType().toString( tenantIdentifier );
```

Reply:
Doing it lika so has the side-effect that now the insert after setting the new tenant id now also doesn't reflect the newly set id, it's just creates another instance with the old tenant id